lib/ecto/type.ex

```diff
@@ -1352,19 +1352,24 @@ defmodule Ecto.Type do
   defp of_base_type?(:date, value), do: Kernel.match?(%Date{}, value)
   defp of_base_type?(_, _), do: false
 
-  defp array([nil | t], fun, true, acc) do
-    array(t, fun, true, [nil | acc])
+  defp array([nil | t], fun, true, {index, acc}) do
```

Comment:
Instead of a tuple, please pass index and acc as separate arguments, it will allocate fewer "objects".

Reply:
Thank you for reviewing this! 😃

That was my first approach but we already have an `array/5` function that takes in a `type` as an extra argument — [see line 1378](https://github.com/elixir-ecto/ecto/blob/v3.11.1/lib/ecto/type.ex#L1378).

I can rename that function to `array_with_type/5` and then do `array_with_index/5` for this one. Wdyt?

Reply:
Renaming is good!